ops/framework.py

```diff
@@ -633,18 +633,18 @@ class SomeObject:
 
         That event may be observed as:
 
-            framework.observe(someobj.something_happened, self.on_something_happened)
-
-        If the method to be called follows the name convention "on_<event name>", it
-        may be omitted from the observe call. That means the above is equivalent to:
-
-            framework.observe(someobj.something_happened, self)
+            framework.observe(someobj.something_happened, self._on_something_happened)
 
+        Raises:
+            RuntimeError: if bound_event or observer are the wrong type.
         """
         if not isinstance(bound_event, BoundEvent):
             raise RuntimeError(
                 'Framework.observe requires a BoundEvent as second parameter, got {}'.format(
                     bound_event))
+        if not isinstance(observer, types.MethodType):
+            raise RuntimeError(
+                'Framework.observe requires a method as third parameter, got {}'.format(observer))
```

Comment:
Can we check for an method on the observer type and give a more helpful error message like:

RuntimeError(
  'Observer methods must now be explicitly provided, please replace observe(..., self) with observe(..., self._on)'
)

IOW, give people the most helpful message we can when breaking compatibility. It shouldn't be hard for them to update their code, but pointing them at how to do so will make it straightforward for them.

Reply:
> check for an method on the observer type

we can, but it might be simpler / better to just raise that one if you pass in something that's a `CharmBase`. WDYT?

(also: IMHO that should be a `TypeError`, not a `RuntimeError` :shrug:

Reply:
I'm flexible about the implementation. Its mostly trying to make it as easy as possible for someone who had working code that said
```
self.framework.observe(self.on.config_changed, self)
```
To respond to the exception they are now getting, and be able to quickly update their code. Whether it is because of checking for CharmBase or checking for a method, I'm not too concerned. Its mostly being able to point them to the quick way to fix the problem.

Reply:
And TypeError vs RuntimeError.... clearly we have a pattern of RuntimeError, given that all the above checks are isinstance that certainly sounds like they should all be TypeError :)